lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -2083,47 +2083,90 @@ void GrandCentralPass::runOnOperation() {
-                              /*includeReplicatedOps=*/true));
-              op->setAttr("comment",
-                          builder.getStringAttr("VCS coverage exclude_file"));
-
-              // Look for any blackboxes instantiated by the companion and mark
-              // them for inclusion in the Grand Central extraction directory.
-              SmallVector<FModuleOp> modules({op});
-              DenseSet<Operation *> bboxes, visited({op});
+
+              // Look for any modules/extmodules _only_ instantiated by the
+              // companion.  If these have no output file attribute, then mark
+              // them as being extracted into the Grand Central directory.
+              SmallVector<Operation *> modules({op});
+              DenseSet<Operation *> visited({op});
+              InstanceGraphNode *companionNode =
+                  instancePaths->instanceGraph.lookup(op);
+
+              LLVM_DEBUG({
+                llvm::dbgs() << "Found companion module: "
+                             << companionNode->getModule().moduleName() << "\n"
+                             << "  submodules exclusively instantiated "
+                                "(including companion):\n";
+              });
               while (!modules.empty()) {
-                auto mod = modules.pop_back_val();
+                auto *mod = modules.pop_back_val();
                 visited.insert(mod);
-                for (auto inst : mod.getOps<InstanceOp>()) {
-                  Operation *sub =
-                      instancePaths->instanceGraph.getReferencedModule(inst);
-                  if (visited.count(sub))
-                    continue;
-                  if (auto subMod = dyn_cast<FModuleOp>(sub)) {
-                    modules.push_back(subMod);
-                    continue;
-                  }
-                  auto subExtMod = cast<FExtModuleOp>(sub);
-                  for (auto anno : AnnotationSet(subExtMod)) {
+
+                DenseSet<FModuleOp> childSet;
+                for (auto path : instancePaths->getAbsolutePaths(mod))
+                  for (InstanceOp instance : path)
+                    childSet.insert(instance->getParentOfType<FModuleOp>());
+
+                // Stop recursiving along this hierarchy if this module is
+                // instantiated by something outside the companion.
+                SmallVector<InstanceRecord *> instances;
+                for (auto *i : instancePaths->instanceGraph.lookup(mod)->uses())
+                  instances.push_back(i);
+
+                // Always modify the output file location of the companion.
+                // Modify the output file location of any modules or external
+                // modules that are instantiated exclusively under the companio.
+                auto isCompanion =
+                    instancePaths->instanceGraph.lookup(mod) == companionNode;
+                if (!isCompanion && !allUnder(instances, companionNode))
+                  continue;
```

Comment:
Just code golfing, and curious if this works. It saves an `InstanceGraph` lookup, simpler syntax on the `SmallVector`:
```
                auto modNode = instancePaths->instanceGraph.lookup(mod);
                SmallVector<InstanceRecord *> instances(modNode->uses());
                if (modNode != companionNode && !allUnder(instances, companionNode))
                  continue;
```
or this one puts the comments nearer to the code and avoids constructing a SmallVector if we are currently the companion, but I dunno
```
                // Always modify the output file location of the companion.
                auto modNode = instancePaths->instanceGraph.lookup(mod);
                if (modNode != companionNode) {
                  // Modify the output file location of any modules or external
                  // modules that are instantiated exclusively under the companion.
                  SmallVector<InstanceRecord *> instances(modNode->uses());
                  if (!allUnder(instances, companionNode))
                    continue;
                }
```

Reply:
Way better! Thanks for taking the time to write up two working suggestions. I'll use the first with a verbose comment about what the check is.